techmatters/hrm
# fix: Twilio channels not being deleted after transcript exported [CHI-2859]

## Description
This PR fixes the bug described in the ticket linked below, by addressing some issues in the `/scheduled-tasks/contact-job-cleanup` task. For some context, this scheduled task is run once every day (triggered by an EventBridge rule). The task itself starts a new hrm container, that:
  - retrieves the contact jobs that are "pending for cleanup". This occurs when 
    - job is marked as "completed" - in the case of exporting transcripts, that is once the transcript has been saved in our own storage.
    - retention period has expired.
  - Depending on the job type, some cleanup processing will be done.
    - In the case of exporting transcripts, the cleanup is deleting the Twilio chat channel.
It has been reported in the bug conversation that jobs where `pending` and never `complete`d. This is because jobs that are "cleaned up" are simply deleted from the DB, so that's why there are no `complete`d cleanup jobs.
The lifecycle of this jobs is as follows:
![image](https://github.com/user-attachments/assets/72adc8d4-786f-48c5-9d23-397bd6be70d6)

However, the current logic for the cleanup has some issues that are addressed in this PR:
- If processing one of the job cleanups failed, the entire process was terminated. This means if there is one broken record in the jobs table, none will be correctly processed until the broken one is purged.
  To solve this issue, I'm introducing a per-job-cleanup `try/catch` block. This will log failing cleanups but won't kill the process, allowing for other job cleanups to continue.
  :bangbang: NOTE: this would be a great place to add some structured logging plus some alerts, to improve monitoring.
- Many of the export transcript job records are in a weird state: they were successfully processed, the transcript was exported and the conversation media record is properly updated, but the `additionalPayload` of the job record was never updated. This is probably some old code misbehaving that caused lots of "broken jobs".
  To solve this issue, I'm not only checking for `additionalPayload.conversationMedia`, but also trying to scan the `ConversationMedias` table using the `contactId` associated the job in the case that the former is absent. If there is conversation media that is a "S3 exported transcript", then we can use this entry as our `conversationMediaId` to check if cleanup is needed.

This PR also contains the changes of https://github.com/techmatters/hrm/pull/685 ([cherry-picked](https://github.com/techmatters/hrm/pull/690)). Original description for the changes:
> This PR fixes yet one more bug around "cleanup jobs".
Basically, before we were only marking accounts as for cleanup check only if they had a `pending` job that was older than `MAX_CLEANUP_JOB_RETENTION_DAYS` (currently set to 1 year).
This PR removes that condition, marking accounts as for cleanup check if they have at least a pending job. We then check, for each account, if there are due due cleanup jobs.



### Checklist
- [x] [Corresponding issue has been opened](https://tech-matters.atlassian.net/browse/CHI-2859)


### Verification steps
This is a hard to verify one, since it's bound to time constrains, the trigger of the scheduled task and some other factors. I am happy to perform some live demos to confirm that this works as intended, but won't ask you to manually check by yourself. You can however scan the development DB and try to look for the existing jobs records, most of which should look fine now (`pending` but not expired  or `not_active`).

### AFTER YOU MERGE

1. Cut a release tag using the GitHub workflow. Wait for it to complete and notify in the #aselo-deploys Slack channel.
2. Comment on the ticket with the release tag version AND any additional instructions required to configure an environment to test the changes.
3. Only then move the ticket into the QA column in JIRA

You are responsible for ensuring the above steps are completed. If you move a ticket into QA without advising what version to test, the QA team will assume the latest tag has the changes. If it does not, the following confusion is on you! :-P
